src/main/java/org/sonar/plugins/jacoco/FileSystemWildcardPatternScanner.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * SonarQube JaCoCo Plugin
+ * Copyright (C) 2018-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.jacoco;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.sonar.api.internal.google.common.annotations.VisibleForTesting;
+import org.sonar.api.utils.WildcardPattern;
+
+public class FileSystemWildcardPatternScanner {
+
+  private static final int SEARCH_MAX_DEPTH = 64;
+
+  private static final String PATH_MATCHER_SPECIAL_CHAR = "*?";
+
+  private final Path knownPathPrefix;
+  @Nullable
+  private final String patternPathSuffix;
+
+  private FileSystemWildcardPatternScanner(String patternPath) {
+    String unixLikePatternPath = toUnixLikePath(patternPath);
+    int specialCharIndex = indexOfMatcherSpecialChar(unixLikePatternPath);
+    if (specialCharIndex == -1) {
+      knownPathPrefix = toFileSystemPath(unixLikePatternPath);
+      patternPathSuffix = null;
+    } else {
+      int knownFolderEnd = unixLikePatternPath.lastIndexOf('/', specialCharIndex);
+      if (knownFolderEnd != -1) {
+        knownPathPrefix = toFileSystemPath(unixLikePatternPath.substring(0, knownFolderEnd + 1));
+        patternPathSuffix = unixLikePatternPath.substring(knownFolderEnd + 1);
+      } else {
+        knownPathPrefix = Paths.get(".");
+        patternPathSuffix = unixLikePatternPath;
+      }
+    }
+  }
+
+  public static FileSystemWildcardPatternScanner of(String patternPath) {
+    return new FileSystemWildcardPatternScanner(patternPath);
+  }
+
+  public void scan(Path baseDirectory, Consumer<Path> pathConsumer, Consumer<String> errorHandler) {
+    scan(baseDirectory, path -> true, pathConsumer, errorHandler);
+  }
+
+  public void scan(Path baseDirectory, Predicate<Path> preFilter, Consumer<Path> pathConsumer, Consumer<String> errorHandler) {
```

Comment:
scan() could return a list and some of the arguments could be hardcoded in the class.